### PR TITLE
DAM Burst Support For InaSAFE

### PR DESCRIPTION
**Pull request by faizalprbw**
### What does it fix?
* Ticket: #we need new classification for dam burst
* Funded by: GEO (faizalprbw)
* Description: Add DAM Break Impact Function and Inundation classes for DAM Break

### Checklist:
- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Add to the changelog in metadata.txt if it's a new feature
- [ ] Unit test for new code added
- [ ] Request someone to review or test your PR

### CODE REVIEW

File: safe/definitions/hazard.py [hunk 1]
```diff
@@ -231,7 +232,59 @@
...
 }
+hazard_dam_break = {
+    'key': 'dam_break',
+    'name': tr('DAM Break'),
+    'description': tr(
+        'A <b>DAM Break</b> is a catastrophic type of failure characterized by '
+        'the sudden, rapid, and uncontrolled release of impounded water '
+        'as a result of structural failures or deficiencies in the dam. '
+        '<b>DAM Break</b> can range from fairly minor to catastrophic, and can '
+        'possibly harm human life and property downstream from the failure. '),
```
Comment (ismailsunni) [inline safe/definitions/hazard.py]: Extra space in the end of sentence.

File: safe/definitions/hazard_classifications.py [hunk 2]
```diff
@@ -657,6 +657,90 @@
...
+            'text': None,
+            'link': None
+        }
+    ],
+    'classes': [
+        {
+            'key': 'inundation_3',
+            'value': 3,
+            'color': red,
+            'name': tr('Inundation Class 3 (Daerah Bahaya 3)'),
```
Comment (ismailsunni) [inline safe/definitions/hazard_classifications.py]: Use English only, the translation shouldn't be here.

File: safe/definitions/hazard_classifications.py [hunk 3]
```diff
@@ -657,6 +657,90 @@
...
+                    'text': None,
+                    'link': None
+                }
+            ]
+        },
+        {
+            'key': 'inundation_2',
+            'value': 2,
+            'color': orange,
+            'name': tr('Inundation Class 2 (Daerah Bahaya 2)'),
```
Comment (ismailsunni) [inline safe/definitions/hazard_classifications.py]: Use English only, the translation shouldn't be here.

File: safe/definitions/hazard_classifications.py [hunk 4]
```diff
@@ -657,6 +657,90 @@
...
+                    'text': None,
+                    'link': None
+                }
+            ]
+        },
+        {
+            'key': 'inundation_1',
+            'value': 1,
+            'color': yellow,
+            'name': tr('Inundation Class 1 (Daerah Bahaya 1)'),
```
Comment (ismailsunni) [inline safe/definitions/hazard_classifications.py]: Use English only, the translation shouldn't be here.

File: resources/img/wizard/keyword-subcategory-dam_break.svg [hunk 5]
```diff
@@ -0,0 +1,100 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (ismailsunni) [inline resources/img/wizard/keyword-subcategory-dam_break.svg]: Compress it to make the size smaller

File: safe/definitions/hazard.py [hunk 6]
```diff
@@ -231,7 +232,59 @@
     'layer_modes': [layer_mode_classified, layer_mode_continuous],
     'disabled_exposures': [exposure_place]
 }
+hazard_dam_break = {
+    'key': 'dam_break',
+    'name': tr('DAM Break'),
```
Comment (Gustry) [inline safe/definitions/hazard.py]: I'm wondering if we should make it lower case. None of our hazard are written with capital letters

File: safe/definitions/hazard_classifications.py [hunk 7]
```diff
@@ -657,6 +657,90 @@
...
 }
 
+inundation_classes = {
+    'key': 'inundation_classes',
+    'name': tr('Inundation classes (DAM Failure)'),
+    'description': tr(
+        'This is a binary classification for an area. The area is either '
+        '<b>wet</b> (affected by flood water) or <b>dry</b> (not affected '
+        'by flood water). This unit does not describe how <b>wet</b> or '
+        '<b>dry</b> an area is.'), # ask
```
Comment (Gustry) [inline safe/definitions/hazard_classifications.py]: The description(wet/dry) doesn't match the classification below with 3,2 and 1 level.
There is a comment saying you need to ask something? ;-)

File: safe/definitions/hazard_classifications.py [hunk 8]
```diff
@@ -657,6 +657,90 @@
...
+    ],
+    'classes': [
+        {
+            'key': 'inundation_3',
+            'value': 3,
+            'color': red,
+            'name': tr('Inundation Class 3 (Daerah Bahaya 3)'),
+            'affected': True, # ask
+            'description': tr('Water is present above ground height.'), # ask
+            'string_defaults': ['wet', '1', 'YES', 'y', 'yes', 'true'], # ask
```
Comment (Gustry) [inline safe/definitions/hazard_classifications.py]: I see there is a few `ask` comments ;-)

File: safe/definitions/hazard_classifications.py [hunk 9]
```diff
@@ -657,6 +657,90 @@
...
+            ]
+        },
+        {
+            'key': 'inundation_1',
+            'value': 1,
+            'color': yellow,
+            'name': tr('Inundation Class 1 (Daerah Bahaya 1)'),
+            'affected': False,
+            'description': tr('No water encountered above ground height.'),
+            'string_defaults': ['dry', '0', 'No', 'n', 'no', 'false'],
```
Comment (Gustry) [inline safe/definitions/hazard_classifications.py]: Description and string_defaults are the same between this class and the previous one.

**Comment by faizalprbw** [comment]
Thanks, i have something to asked but then i was confusing by other project
so i forgot to ask those until yesterday world bank ask me for dam break
workshop..
Actually i'been removed those in my local repo.. but if you could help me
to explain what is that dictionary suppose to be? should i put that in my
hazard classification?

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Thu, Nov 30, 2017 at 2:19 PM, Etienne Trimaille <notifications@github.com

File: safe/definitions/hazard.py (hunk 6 above)
Comment (faizalprbw) [reply safe/definitions/hazard.py]: Thanks, i have something to asked but then i was confusing by other project so i forgot to ask those.. 
Actually i've been removed those in my local repo.. but if you could help me to explain what is that dictionary suppose to be? should i put that in my hazard classification?

Comment (faizalprbw) [reply safe/definitions/hazard.py]: hhm what about this?
 hazard_flood = {
    'key': 'flood',
    'name': tr('Flood')

Comment (faizalprbw) [reply safe/definitions/hazard.py]: oh i see..

Comment (ismailsunni) [reply safe/definitions/hazard.py]: `name` is the human friendly name, while `key` is the computer friendly one.
Does `DAM` means `Bendungan` in Bahasa? If so, you should write it as `Dam`. Or is it a abbreviation?

I think you should also explain what is DAM in this case in the description.

File: resources/img/wizard/keyword-subcategory-dam_break.svg (hunk 5 above)
Comment (Gustry) [reply resources/img/wizard/keyword-subcategory-dam_break.svg]: This SVG is not in the correct folder. It must be in **safe**/resources/img/wizard/keyword-subcategory-dam_break.svg
Then Travis will be happy

**Comment by Gustry** [comment]
* You are also pushing QPT to `resources/qgis-composer-templates/GEO_DAM_BREAK_TEMP/inasafe-map-report-portrait.qpt`. This the old directory. It's not in SAFE.
* Why did you update the provided QPT? We would need to check these.
* What are the new QPT? Do we need them in InaSAFE core? 'GEO_DAM_BREAK_TEMP' ?

Well done, Travis will be happy if you update the svg path

File: resources/img/wizard/keyword-subcategory-dam_break.svg (hunk 5 above)
Comment (faizalprbw) [reply resources/img/wizard/keyword-subcategory-dam_break.svg]: i see.. here i push my update..

**Comment by Gustry** [comment]
FYI, Travis won't run if the PR has some conflicts. You need to fix them first.
Do you need these update QPT? What did you change inside or is-it specific to your project?

If it's specific to our project, you can call them with different name, InaSAFE will still be able to pick them in the print dialog (correct me if I'm wrong @myarjunar )

**Comment by faizalprbw** [comment]
No, i made GEO_DAM_BREAK_TEMP just in case to autolink the logo 1 month ago, and i forget to remove it.. But then i remove it, after that i merge it with current develop branch and got these conflict.. No, i'm trying to fix some conflict..:

File: safe/definitions/hazard.py [hunk 10]
```diff
@@ -231,7 +232,61 @@
...
 }
+hazard_dam_break = {
+    'key': 'dam_break',
+    'name': tr('Dam Break'),
+    'description': tr(
+        'A <b>Dam Break</b> is a catastrophic type of failure '
+        'characterized by the sudden, rapid, and uncontrolled '
+        'release of impounded water as a result of structural '
+        'failures or deficiencies in the dam. '
+        '<b>DAM Break</b> can range from fairly minor to '
```
Comment (Gustry) [inline safe/definitions/hazard.py]: Still one DAM with capital letters

File: safe/definitions/hazard.py [hunk 11]
```diff
@@ -231,7 +232,61 @@
...
+        'catastrophic, and can possibly harm human life and property '
+        'downstream from the failure. '),
```
Comment (Gustry) [inline safe/definitions/hazard.py]: Final space not needed as it's end of the string

File: safe/definitions/hazard_classifications.py [hunk 12]
```diff
@@ -657,6 +657,90 @@
     'classification_unit': tr('hazard zone')
 }
 
+inundation_classes = {
+    'key': 'inundation_classes',
+    'name': tr('Inundation classes (Dam Failure)'),
```
Comment (Gustry) [inline safe/definitions/hazard_classifications.py]: If the classification is specific to dam break, maybe you should rename the variable `inundation_dam_class` L660 and key also. You can rename the name to 'Inundation classes', this class will be displayed only for the dam failure hazard already so not need to specify it again in the name here.

File: safe/definitions/hazard_classifications.py [hunk 13]
```diff
@@ -657,6 +657,90 @@
...
+    'description': tr(
+        'This is a binary classification for an area. The area is either '
+        '<b>wet</b> (affected by flood water) or <b>dry</b> (not affected '
+        'by flood water). This unit does not describe how <b>wet</b> or '
+        '<b>dry</b> an area is.'),
```
Comment (Gustry) [inline safe/definitions/hazard_classifications.py]: The description is not correct. It refers actually to a binary classification, but it seems you have 3 classes below.

File: safe/definitions/hazard_classifications.py [hunk 14]
```diff
@@ -657,6 +657,90 @@
...
+                }
+            ]
+        },
+        {
+            'key': 'inundation_1',
+            'value': 1,
+            'color': yellow,
+            'name': tr('Inundation Class 1'),
+            'affected': False,
+            'description': tr('No water encountered above ground height.'),
```
Comment (Gustry) [inline safe/definitions/hazard_classifications.py]: Same as line 701

File: safe/definitions/hazard_classifications.py [hunk 15]
```diff
@@ -657,6 +657,90 @@
...
+            'string_defaults': ['dry', '0', 'No', 'n', 'no', 'false'],
```
Comment (Gustry) [inline safe/definitions/hazard_classifications.py]: Same as line 702. I guess it's a copy/paste before ;-)

**Comment by ismailsunni** [comment]
@faizalprbw don't forget to update your PR with the current develop branch.

**Comment by Gustry** [comment]
It's already updated @ismailsunni no more conflict with the PR. We just can't rebase because individual commits have conflicts with QPT, but not the PR itself. We can squash them at the end.

Only one comment left

**Comment by ismailsunni** [comment]
Ah I see, sorry, my default option is rebase, so I didn't realize if it's green if we merge/squash it.

**Comment by faizalprbw** [comment]
what should i do to merge it? 
is there anything i have to do?

**Comment by timlinux** [comment]
Thank you very much for the new hazard class. I think it would be good to also add notes and checklist items that are specific to dam break? Also can you you please share any training materials you developed for this features so that we can incorporate into our standard documentation?

**Comment by faizalprbw** [comment]
what does  notes and checklist items means? for training materials, i'll share those on in google drive..

**Comment by faizalprbw** [comment]
https://drive.google.com/open?id=1PjYJvl46Qbtqi4_iCQi3zUJ01_XPekSp
Here are the data for Dam Break

File: safe/definitions/hazard_classifications.py (hunk 14 above)
Comment (Gustry) [reply safe/definitions/hazard_classifications.py]: Can you fix that one, line 701 and 720 are the same. It can't be the same description.
Thanks

**Comment by Gustry** [comment]
@timlinux I checked, most of notes and checklists are provided by the exposure. For hazard, we don't have too much.

@faizalprbw Actions and notes are displayed in the HTML and PDF reports.
You can have some examples of actions here for http://manual.inasafe.org/en/index.html#Land-cover in 7.1.1.1.3. They provide a checklist for the DRM what to check/do in case of emergency or in case of an assessment.

**Comment by faizalprbw** [comment]
Are the action list and general note depend on the combination between impact function and exposure type too? or they just defining based on exposure?

**Comment by Gustry** [comment]
Actions and notes can be defined at different levels:
* If they are exposure specific only: in `exposure.py`
* If hazard specific only: in `hazard.py` the file you edited in this PR by adding a new hazard
* If actions and notes are a combination of hazard/exposure: in 'hazard_exposure_specifications.py'

We all display them in reports, in the dock after the analysis is completed.

**Comment by faizalprbw** [comment]
oh i missed it, i forgot to ask to them about it. All they had suggested was just add a new impact function which is as same as flood and also inundation class. So i thought that action list and note will be similar to flood hazard..

**Comment by Gustry** [comment]
ok, no problem. I'm going to merge it. We can check that later if needed. Thanks !

**Comment by faizalprbw** [comment]
Thanks, just ask me if there are something i have to work wit, edit, or added for Dam Break

**Pull request merged**